debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java

```diff
@@ -139,110 +150,213 @@ public String getName() {
+        if (m.group("mod") != null) {
+            typeModifiers = TYPEMOD_PATTERN.split(m.group("mod"));  // TODO: use for decimal/etc precision
+        }
+        boolean isArray = (m.group("array") != null);
+
+        if (baseType.startsWith("_")) {
+            // old-style type specifiers use an _ prefix for arrays
+            // e.g. int4[] would be "_int4"
+            baseType = baseType.substring(1);
+            fullType = fullType.substring(1);
+            isArray = true;
+        }
+
+        if (rawValue.isNull()) {
+            // nulls are null
+            return null;
+        }
+
+        if (isArray) {
+            try {
+                final String dataString = rawValue.asString();
+                PgArray arrayData = new PgArray(connection.get(), connection.get().getTypeInfo().getPGArrayType(fullType), dataString);
+                Object deserializedArray = arrayData.getArray();
+                // TODO: what types are these? Shouldn't they pass through this function again?
+                return Arrays.asList((Object[])deserializedArray);
+            }
+            catch (SQLException e) {
+                LOGGER.warn("Unexpected exception trying to process PgArray ({}) column '{}', {}", columnType, columnName, e);
+            }
+            return null;
+        }
+
+        switch (baseType) {
+            // include all types from https://www.postgresql.org/docs/current/static/datatype.html#DATATYPE-TABLE
+            // plus aliases from the shorter names produced by older wal2json
+            case "boolean":
             case "bool":
-                return rawValue.isNotNull() ? rawValue.asBoolean() : null;
-            case "int2":
+                return rawValue.asBoolean();
+
+            case "integer":
+            case "int":
             case "int4":
-                return rawValue.isNotNull() ? rawValue.asInteger() : null;
-            case "int8":
+            case "smallint":
+            case "int2":
+            case "smallserial":
+            case "serial":
+            case "serial2":
+            case "serial4":
             case "oid":
-                return rawValue.isNotNull() ? rawValue.asLong() : null;
+                return rawValue.asInteger();
+
+            case "bigint":
+            case "bigserial":
+            case "int8":
+                return rawValue.asLong();
+
+            case "real":
             case "float4":
-                return rawValue.isNotNull() ? rawValue.asFloat() : null;
+                return rawValue.asFloat();
+
+            case "double precision":
             case "float8":
-                return rawValue.isNotNull() ? rawValue.asDouble() : null;
+                return rawValue.asDouble();
+
             case "numeric":
-                return rawValue.isNotNull() ? rawValue.asDouble() : null;
+            case "decimal":
+                // TODO: When wal2json supports decimal/numeric, return via .asBigDecimal()
```

Comment:
That seems to be the case now?

Reply:
Yeah, it does, I verified the toString() representation is "right", but I haven't figured out the JSON-parsing side in Java yet.

See numeric comments above.